test/linalg/dense.jl

```diff
@@ -582,6 +582,25 @@ end
     end
 end
 
+@testset "issue #23366 (Int Matrix to Int power)" begin
+    @testset "Tests for $elty" for elty in (Int128, Int16, Int32, Int64, Int8,
+                                            UInt128, UInt16, UInt32, UInt64, UInt8,
+                                            BigInt)
+        @test_throws DomainError elty[1 1;1 0]^-2
+        @test elty[1 1;1 0]^2 == elty[2 1;1 1]
+        I_ = elty[1 0;0 1]
+        @test I_^-1 == I_
+        if !(elty<:Unsigned)
+            @test (-I_)^-1 == -I_
+            @test (-I_)^-2 == I_
+        end
+        for elty2 = (Int64, BigInt)
+            TT = Base.promote_op(^, elty, elty2)
+            @test Base.return_types(^, (Array{elty,2}, elty2)) == [Array{TT,2}]
```

Comment:
The test isn't redundant because there's an additional issue going on that I discovered while fixing this issue. So far `[1 1;1 0]^big(10000)` returned `Matrix{Int64}` (and negative values) rather than promoting to `Matrix{BigInt}`. This was inconsistent with `2^big(10000)`. I fixed this issue as well.